### ISSUE TITLE
Do I have to perform optimistic concurrency check by hand when a commit is saved ?

### ISSUE THREAD

**Issue by EnricoMassoneDeltatre** [NEventStore/NEventStore#477]
Hi,

this is a beginner question because I don't have any experience in using NEventStore and I'm giving it a try.

The point of this question is related to the concept of optimistic concurrency check as envisioned by Greg Young [in this document](https://cqrs.files.wordpress.com/2010/11/cqrs_documents.pdf) for which a practical example is given [here](https://github.com/gregoryyoung/m-r/blob/master/SimpleCQRS/EventStore.cs).

So, in my application I have the following interface which is the event store abstraction that I'm going to use in the repository implementation: 

```
public interface IEventStore
    {
        void SaveEvents(Guid aggregateId, IEnumerable<Event> events, int expectedVersion);
        List<Event> GetEventsForAggregate(Guid aggregateId);
    }
```

My goal is providing an implementation of `IEventStore` by using the NEventStore library. This is a naive implementation, where I perform a kind of *poor man optimistic concurrency check*: 

```
public class EventStore : IEventStore 
{
	private readonly IEventStream stream;

	public EventStore(IEventStream stream)
	{
		this.stream = stream ?? throw new ArgumentNullException(nameof(stream));
	}
	
	public List<Event> GetEventsForAggregate(Guid aggregateId)
	{
		// implementation omitted because I'm only interested in understanding how to imeplement save method right now...
	}
	
	public void SaveEvents(Guid aggregateId, IEnumerable<Event> events, int expectedVersion)
	{
		using (var stream = this.store.OpenStream(aggregateId, 0, int.MaxValue))
                {

		  // here, by following Greg Young's paper, I should perform the optimistic concurrency check...
		  int currentAggregateRevision = stream.StreamRevision;
		  bool isNewStream = currentAggregateRevision == 0;
		  
		  if(!isNewStream && expectedVersion != currentAggregateRevision)
		  {
			// DANGER: optimistic concurrency check failed !
			throw new ConcurrencyException("The guy that issued the command did not work on the latest version of the aggregate. We cannot commit these events.")
		  }
		
                  foreach(var @event in events)
		  {
			stream.Add(new EventMessage { Body = @event });  
		  }

                  stream.CommitChanges(Guid.NewGuid()); // Is there a best practice to generate a commit id ? Is it ok to use a new guid ?
               }
	}
}
```

Is this a correct approach to implement greg young's `IEventStore` interface ? 

Do I have to manually implement the optimistic concurrency check as I did above or the NEventStore library is capable to do it on my behalf ? 

Thanks for helping

Enrico

**Comment by gregoryyoung** [comment]
This isn't safe unless you are also locking the stream. What if there are two threads/nodes doing this at the same time?

**Comment by EnricoMassoneDeltatre** [comment]
> This isn't safe unless you are also locking the stream. What if there are two threads/nodes doing this at the same time?

Hi @gregoryyoung,

so we are discussing about two users performing an operation on the same stream at the same time. You are saying that, even if both of them are working on the latest aggregate version, there is still a "race condition" between their commands. 

By looking at the xml documentation for the method `IEventStream.CommitChanges(Guid commitId)` I can spot an exception of type ConcurrencyException. I think it's the same kind of concurrency exception to which [this other issue](https://github.com/NEventStore/NEventStore/issues/472) refers to.

So, to summarize, my understanding is that the NEventStore library **is capable of locking the event stream for me**, so that when I commit some events (like in the code above) I only have to be concerned with the events I'm trying to commit and i can safely ignore concurrent changes on the same event stream.

Is my understanding correct ? Am I missing anything ? 

Thanks for helping

**Comment by fschmied** [comment]
> Do I have to manually implement the optimistic concurrency check as I did above or the NEventStore library is capable to do it on my behalf ?

NEventStore will throw on `Save` if the version the stream was originally loaded with doesn't match the current version of the stream when it's saved (optimistic concurrency control). So, as an alternative, you could also just load the stream up to the version you're interested in (instead of passing `int.MaxValue`), append the new events, then call `Save` and see what happens.

If I'm not missing anything, checking the stream version beforehand therefore is probably an optimization only. I'm also doing it in my project, though.

Looking at https://github.com/NEventStore/NEventStore.Domain (formerly known as "CommonDomain"), which is a kind of example implementation of a higher-level abstraction around NEventStore, they seem to rely only on the `Save` check. (I'm not using that code as it contains a lot of stuff I don't (currently) need, such as snapshot support. Also I'm not sure I like all that caching code, which seems a bit fishy to me regarding thread-safety and concurrency checks.)

> Is this a correct approach to implement greg young's IEventStore interface ?

Your concurrency check looks mostly the same as mine, however it differs in the way you're handling new streams: `if(!isNewStream && expectedVersion != currentAggregateRevision)`. Why wouldn't you throw the exception if the `expectedVersion` is, say , 12, but NES says it's a new stream?

In my code, `expectedVersion` is nullable, by the way. I quite often have commands that don't care about concurrent changes, so they don't specify an expected version at all. If `Save` throws a `ConcurrencyException` in such a case, I'll simply retry the operation.

> So, to summarize, my understanding is that the NEventStore library is capable of locking the event stream for me, so that when I commit some events (like in the code above) I only have to be concerned with the events I'm trying to commit and i can safely ignore concurrent changes on the same event stream.

It's optimistic concurrency control, so you can't really "ignore concurrent changes" - you'll need to think about how you'll deal with the exception... But apart from that, yes, NEventStore will detect the concurrent change and not allow the later `Save` to succeed.

**Comment by EnricoMassoneDeltatre** [comment]
Hi @fschmied, 

thank you for the in depth explanation. 

In the [simple CQRS project](https://github.com/gregoryyoung/m-r) Greg uses a convention for which when a new aggregate is created (so you expected that the stream doesn't exist in the event store) the variable `expectedVersion` has a value of `-1`.  
This basically means that you track the aggregate version with a zero-based index (the first event you apply gets the aggregate to revision `0`, the second one to revision `1` and so on...). 

The NEventStore approach seems to be one-based index for aggregate revision (I only tried the mongodb persistence, but I guess it's the same for all the supported storage), so I could use the convention of having `expectedVersion==0` for newly created aggregates. In that case I can also use the method `IStoreEvent.CreateStream(Guid streamId)` to open the stream, because the call to `IStoreEvent.CommitChanges(Guid commitId)` throws a `ConcurrencyException` if a stream having the same Id already exists (so we have a safe guard against creating the same stream twice).  

Based on your suggestion, I plan to change my code in the following way: 

```
public class EventStore : IEventStore 
{
	private readonly IEventStream stream;

	public EventStore(IEventStream stream)
	{
		this.stream = stream ?? throw new ArgumentNullException(nameof(stream));
	}
	
	public List<Event> GetEventsForAggregate(Guid aggregateId)
	{
		// implementation omitted because I'm only interested in understanding how to imeplement save method right now...
	}
	
	public void SaveEvents(Guid aggregateId, IEnumerable<Event> events, int expectedVersion)
	{
		if(expectedVersion < 0)
		{
		  throws new ArgumentOutOfRangeException(nameof(expectedVersion));
		}

		bool isNewStream = expectedVersion == 0;

		if(isNewStream) {

		  using (var stream = this.store.CreateStream(aggregateId))
		  {
			foreach(var @event in events)
			{
				stream.Add(new EventMessage { Body = @event });  
			}

			stream.CommitChanges(Guid.NewGuid()); // throws ConcurrencyException if an aggregate with the same id already exists... 
		  }

		} else {

		  using (var stream = this.store.OpenStream(aggregateId, 0, expectedVersion))
		  {
			foreach(var @event in events)
			{
				stream.Add(new EventMessage { Body = @event });  
			}

			stream.CommitChanges(Guid.NewGuid()); // throws ConcurrencyException in case of concurrency issues...
		  }

		}
	}
}
```
In my opinion this is much cleaner and expressive than the previous version.

**Comment by fschmied** [comment]
Looks good to me as far as I understand the NEventStore APIs.

PS: I edited [my previous comment](https://github.com/NEventStore/NEventStore/issues/477#issuecomment-498043036) a little bit and added a link directly to NEventStore.Domain's EventStoreRepository.

**Comment by EnricoMassoneDeltatre** [comment]
> Looks good to me as far as I understand the NEventStore APIs.
> 
> PS: I edited [my previous comment](https://github.com/NEventStore/NEventStore/issues/477#issuecomment-498043036) a little bit to fix a mistake (I had mixed up the terms `Save` and `CommitChanges`) and added a link directly to NEventStore.Domain's EventStoreRepository.

Thank you very much for your comments.
